### PR TITLE
[21.2] Shader Graph: Added Matrix Swizzle Node

### PR DESCRIPTION
**Pull request by tracychen1234**
---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?
Added the Matrix Swzzile node, updated AddSlot() in AbstractMaterialNode.cs. 
image
When users change the dropdown, the input boxes will be grayed out according to the selected output size. The node will only allow matrix as input but can output both matrix and vector. 

---
### Testing status

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 
Put node in the graph and made sure it's output the correct size according to the selected enum list. Tested out different matrix output nodes as its input. 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)
In progress.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Comment by alindmanUnity** [comment]
working with tracy to add tests to automated suite

**Comment by LandonTownsendUnity** [comment]
Can you add some code reviewers? I'm curious about the halo effect of some of the changed files.

**Comment by LandonTownsendUnity** [comment]
I thought the plan was to have 1 to 1 conversion values be the "default"... but it looks like it's all 0s?
image
I thought that it was going to be 00 01 02 03, 10 11 12 13, etc. Was that plan changed?

**Review by ghost — CHANGES_REQUESTED** [review]
Shouldn't 04 be an invalid index?
image

**Review by ghost — CHANGES_REQUESTED** [review]
image
Deleting the contents of a text box and typing anything that isn't a number (such as a letter or a space) permanently corrupts the box, making it impossible to type anything in it but xx

**Comment by tracychen1234** [comment]
Fixing it now!

**Comment by LandonTownsendUnity** [comment]
I also think that the default input should be a 4 x 4 identity matrix, not a 2x2, so that the user can start filling it out before plugging the matrix in (or have it filled in by default) without getting an out of range error

**Comment by LandonTownsendUnity** [comment]
The default input is still a 2x2 matrix, meaning you get a "Validation: invalid index!" right out of the gate when you create one.

**Review by ghost — CHANGES_REQUESTED** [review]
Something is wrong with the row column order here.

image
I'm using this shadergraph to test.
image
image
For some reason, the starting values, which are supposed to give me the matrix that I started with, aren't working correctly.

File: com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+                inputIndices.SetRow(r, indices_row);
+                AreIndiciesValid = true;
+            }
+
+            //build shader string
+            string real_outputValue = "";
+            for (var r = 0; r < outputRowCount; r++)
+            {
+                string outputValue = "";
+                Vector4 indecies = inputIndices.GetRow(r);
```
Comment (LandonTownsendUnity) [inline com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs:337]: You may want to correct the spelling on this variable

File: com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+            }
+            else
+            {
+                AreIndiciesValid = false;
+                ValidateNode();
+            }
+            return new Vector4(0, 0, 0, 0);
+        }
+
+        private bool AreIndiciesValid = true;
```
Comment (LandonTownsendUnity) [inline com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs:208]: You shouldn't put member variables in between functions like this, this should be at the top with the other member variables. Also, the spelling needs to be fixed.

File: com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+            AreIndicesValid = true;
+            //Check indices sizes
+            for (int r = 0; r< 4; r++)
+            {
+                Vector4 indices_row = inputIndices.GetRow(r);
+                for(int c = 0; c<4; c++)
+                {
+                    float cell = inputIndices.GetRow(r)[c];
+                    int[] cell_lst = getIndex(cell);
+                    if (cell_lst[0] > concreteRowCount-1 || cell_lst[1] > concreteRowCount - 1)
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Nodes/Math/Matrix/MatrixSwizzleNode.cs:321]: x > y-1
better to use:
x >= y
when they are both expected to be integers

**Review by cdxntchou — CHANGES_REQUESTED** [review]
I think the big thing to address here is that you've got too many ways to represent the swizzle values, which is complicating the code a lot.  You've got the combined 8-character string, the exploded 8-character array, the individual strings per field, and the floating point combined row.col value.  I think it would simplify things a lot if we just had a "SwizzleRow" class that contained four separate strings, one for each column, and used that as your base data structure in place of the combined 8-character string.

**Review by ValGrimm-U3D — CHANGES_REQUESTED** [review]
Pasting over revised text for Matrix-Swizzle-Node.md (which was fixed in the Internal repo)

Reorders the elements of the input vector as you specify in the input mask. The length of the input mask determines the dimensions of the output vector.
The error "Invalid Mask" indicates an input mask value which includes one or more channels that do not exist in the input vector.

**Comment by tracychen1234** [comment]
Closed the PR because it's not on our 21.2 plan.

**Pull request closed without merging**